MetadataExtractor/IO/RandomAccessStream.cs

```diff
@@ -0,0 +1,561 @@
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            // TODO: good spot to try Span<T>
+            if (p_isStreamFinished && p_chunks.Count == 1)
+            {
+                Array.Copy(p_chunks[0], (int)index, buffer, 0, count);
+                return count;
+            }
+
+            var remaining = count;                      // how many bytes are requested
+            var fromOffset = (int)index;
+            var toIndex = offset > 0 ? offset : 0;
+            while (remaining != 0)
+            {
+                var fromChunkIndex = fromOffset / p_chunkLength;     // chunk integer key
+                var fromInnerIndex = fromOffset % p_chunkLength;     // index inside the chunk to start reading
+                var length = Math.Min(remaining, p_chunkLength - fromInnerIndex);
+                var chunk = p_chunks[fromChunkIndex];
+                Array.Copy(chunk, fromInnerIndex, buffer, toIndex, length);
+                remaining -= length;
+                fromOffset += length;
+                toIndex += length;
+            }
+
+            return toIndex - offset;
+        }
+
+        /// <summary>Returns an unsigned byte at an index in the sequence.</summary>
+        /// <returns>the 8 bit int value, between 0 and 255</returns>
+        /// <param name="index">position within the data buffer to read byte</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        public byte GetByte(long index, bool isSequential)
+        {
+            ValidateIndex(index, 1, isSequential);
+
+            // This bypasses a lot of checks particularly when the input was a byte[]
+            if (p_isStreamFinished && p_chunks.Count == 1)
+                return p_chunks[0][index];
+
+            var chunkIndex = index / p_chunkLength;
+            var innerIndex = index % p_chunkLength;
+
+            if (p_chunks.ContainsKey(chunkIndex))
+                return p_chunks[chunkIndex][innerIndex];
+            else
+                return unchecked((byte)-1);
+        }
+
+        /// <summary>Returns an unsigned 16-bit int calculated from the next two bytes of the sequence.</summary>
+        /// <returns>the 16 bit int value, between 0x0000 and 0xFFFF</returns>
+        /// <param name="index">position within the data buffer to read first byte</param>
+        /// <param name="IsMotorolaByteOrder">byte order for returning the result</param>
+        /// <param name="isSequential">flag indicating if caller is using sequential access</param>
+        /// <exception cref="BufferBoundsException"/>
+        /// <exception cref="IOException"/>
+        public ushort GetUInt16(long index, bool IsMotorolaByteOrder, bool isSequential)
+        {
+            var bytes = new byte[2];
```

Comment:
These `Get*` methods are called a lot and should not be allocating temporary arrays.

Reply:
Ok. I used Read to make the access consistent, but I'll try to write another internal method that reads the buffer Dictionary directly instead. Might take a little testing but seems possible.

Reply:
I went back and used GetByte, which is actually the original solution. I changed it later thinking Read was more efficient, but the extra byte arrays are probably too much. However, to avoid validating all the time, GetByte has a new overload to turn off its own validation call.

Reply:
Sounds good. At some point in the future I think we can consider dropping .NET Framework support and going all in on `Span<>` and friends, which should make all this stuff much nicer, and faster.

Reply:
(Unresolving to continue discussion.)

Could these `Get*` methods (other than `Get(byte[]...`) be moved to `ReaderInfo`? If so each `ReaderInfo` could have its own `byte[8]` as a field which it would re-use across calls. I think it would be reasonable to ensure a given `ReaderInfo` could not be used in parallel. I suspect filling a `byte[]` once with N bytes would be faster than reading N bytes one at a time (again, we need some benchmarks to verify these hypotheses).